internal: Add Nix flake and README.md

laxsjo opened the pull request:
This PR adds a flake.nix with a development shell, a formatter for the Nix code, and checks for the formatter and parser.

It also adds a README.md with some basic notes about how to enter the development environment and generate and test the parser.

I decided to only add the tree-sitter CLI and the dependencies required to install the 'node_modules/' directory into the development shell since I don't know of any others that you would need. Have I missed any that you're using @cortex?

I also took the oportunity to add myself to all of the contributor/author fields.

laxsjo committed:
chore: Ignore .vscode

Added the .vscode directory to the '.gitignore' file.

laxsjo committed:
internal: Add flake.nix

This commit adds a flake.nix with a dev-shell and formatting configured.

I decided that it would be a good idea to enable formatting for our nix
files immediately, so that we don't need to worry about reformatting
existing Nix code down the line. I didn't add any other formatters for
now however.

laxsjo committed:
test: Add Tree-sitter checks to flake.nix

Added two new checks to the Nix flake, checking that
`tree-sitter generate` has been ran, and that the tests in the 'test/'
directory pass.

Also note that it seems like regenerating the parser changed a some of
the files under `src/`. I guess those were due to changes in 'parser.js'
that hadn't been regenerated in an earlier commit and/or that the version of the
`tree-sitter` CLI I used did things slightly differently.

laxsjo committed:
docs: Add README.md

Added a basic README with notes about the development environment and
instruction for how to generate and test the parser.

We should add more details in the future, especially with instructions
for how to use it in other projects.

cortex commented on flake.nix:
```diff
@@ -0,0 +1,100 @@
+                  pkgs.nodejs_23
+                ];
+              }
+              ''
+                # Create source directory with a git tree.
+                set -e
+                cp -r ${self} source/
+                chmod -R a+w source/
+                cd source/
+                export HOME=$TMPDIR
+                cat > $HOME/.gitconfig <<EOF
+                [user]
+                  name = Nix
+                  email = nix@localhost
+                [init]
+                  defaultBranch = main
+                EOF
+                git init
+                git add .
+                git commit -m init --quiet
+
+                tree-sitter generate
+
+                # Check that nothing changed
+                git status
+                if ! git --no-pager diff --exit-code; then
+                  echo "Generated parser not up to date with 'parser.js'. Please run the following:"
+                  echo "  tree-sitter generate"
+                  exit 1
+                fi
+
+                touch $out
+              '';
```
This is a fun way to use git :) Maybe this could be made simpler using diff --recursive instead?

laxsjo replied:
Perhaps, I copied it from the check used by treefmt-nix. You can read it here: https://github.com/numtide/treefmt-nix/blob/adc195eef5da3606891cedf80c0d9ce2d3190808/module-options.nix#L237-L272

But using `diff` does sound a lot more elegant, let me try it!

laxsjo replied:
Forced pushed a version which uses `diff` instead. Thanks for the suggestion!

laxsjo committed:
build: Add Cargo.lock

This seems to be generated when you run `tree-sitter generate`. As it
locks the build inputs it should be committed to the repo.

cortex commented on flake.nix:
```diff
@@ -0,0 +1,100 @@
+              }
+              ''
+                # Create source directory.
+                set -e
+                cp -r ${self} source/
+                chmod -R a+w source/
+                cd source/
+                export HOME=$TMPDIR
+
+                tree-sitter test
```
Does this mean that tree-sitter test cannot operate on a readonly directory? That's a bit silly

laxsjo replied:
Honestly not sure, I just copy pasted that part from another project I found. Let me check if it's actually necessary :P

laxsjo replied:
Update: It was of course completely fine to leave the directory as readonly. I force pushed a version with it removed.

laxsjo committed:
chore: Add Rasmus Söderhielm as contributor

Added myself as a contributor to all of the places which lists authors
or contributors.

cortex reviewed (approved):
Very nice thanks for this!

laxsjo commented:
I think this is ready to merge now unless you have anything else. :)
